libvirt/utils_volume.go

```diff
@@ -179,18 +183,30 @@ func (i *httpImage) Import(copier func(io.Reader) error, vol libvirtxml.StorageV
 	if vol.Target.Timestamps != nil && vol.Target.Timestamps.Mtime != "" {
 		req.Header.Set("If-Modified-Since", timeFromEpoch(vol.Target.Timestamps.Mtime).UTC().Format(http.TimeFormat))
 	}
-	response, err := client.Do(req)
 
-	if err != nil {
-		return fmt.Errorf("Error while downloading %s: %s", i.url.String(), err)
-	}
+	var response *http.Response
+	for retryCount := 0; retryCount < maxHTTPRetries; retryCount++ {
+		response, err = client.Do(req)
+		if err != nil {
+			return fmt.Errorf("Error while downloading %s: %v", i.url.String(), err)
+		}
+		defer response.Body.Close()
 
-	defer response.Body.Close()
-	if response.StatusCode == http.StatusNotModified {
-		return nil
+		log.Printf("[DEBUG]: url resp status code %s (retry #%d)\n", response.Status, retryCount)
+		if response.StatusCode == http.StatusNotModified {
+			return nil
+		} else if response.StatusCode == http.StatusOK {
+			return copier(response.Body)
+		} else if response.StatusCode < 500 {
+			break
+		} else {
+			// retry
+			if retryCount < maxHTTPRetries {
+				time.Sleep(retryWaitSec * time.Second)
+			}
+		}
 	}
-	log.Printf("[DEBUG]: url resp status code %s\n", response.Status)
-	return copier(response.Body)
+	return fmt.Errorf("Error while downloading %s: %v", i.url.String(), response)
```

Comment:
we could maybe say `unexpected Error while downloading ..` and differentiate the log,  because we are returning when we don't catch the error

Reply:
I don't think any of these errors is "expected".